src/common/interceptors/mask-sensitive-data.interceptor.ts

```diff
@@ -0,0 +1,99 @@
+import {
+  CallHandler,
+  ExecutionContext,
+  Injectable,
+  Module,
+  NestInterceptor,
+} from "@nestjs/common";
+import { APP_INTERCEPTOR } from "@nestjs/core";
+import { instanceToPlain } from "class-transformer";
+import { isEmail } from "class-validator";
+import { map, Observable } from "rxjs";
+
+@Injectable()
+class MaskSensitiveDataInterceptor implements NestInterceptor {
+  private readonly keysToMask = ["orcidOfOwner"];
+
+  private maskSensitiveData<T>(data: T, ownEmail: string): T {
+    if (Array.isArray(data)) {
+      let allArrays = true;
+      const maskedData = data.map((item) => {
+        if (typeof item === "string" && this.isToMaskEmail(item, ownEmail)) {
+          return this.maskValue();
+        }
+        allArrays = false;
+        return this.maskSensitiveData(item, ownEmail);
+      });
+      return ((allArrays && [...new Set(maskedData)]) || maskedData) as T;
+    }
+
+    if (!this.isPlainObject(data)) return data as T;
+    const result: Record<string, unknown> = {};
+    for (const [key, value] of Object.entries(data)) {
+      if (
+        this.keysToMask.includes(key) ||
+        (typeof value === "string" && this.isToMaskEmail(value, ownEmail))
+      ) {
+        result[key] = this.maskValue();
+        continue;
+      }
+      result[key] = this.maskSensitiveData(value, ownEmail);
+    }
+    return result as T;
+  }
+
+  private isPlainObject(input: unknown): input is Record<string, unknown> {
+    return (
+      typeof input === "object" &&
+      input !== null &&
+      !Array.isArray(input) &&
+      Object.getPrototypeOf(input) === Object.prototype
+    );
+  }
+
+  private isToMaskEmail(value: string, ownEmail: string) {
+    return isEmail(value) && value !== ownEmail;
+  }
+
+  private maskValue(): string {
+    return "";
+  }
```

Comment:
Is it allowed to return partial information under the regulation instead of just returning empty string? e.g, `ju*****@ess.eu`

Reply:
I don't have a strong opinion on the value. Something like `gdpr_hidden`,  `not available(gdpr)` would work maybe?

Reply:
what if I use `*****`?